docs/design/ovs-pipeline.md

```diff
@@ -1,1197 +1,1938 @@
 metadata:
-  name: cnp0
+  name: web-app-db-network-policy
 spec:
-  priority: 10
-  tier: application # defaults to application tier if not specified
+  priority: 5
+  tier: application
   appliedTo:
     - podSelector:
         matchLabels:
-          app: server
+          app: web
   ingress:
-    - action: Drop
+    - action: Allow
       from:
         - podSelector:
             matchLabels:
-              app: notClient
+              app: client
       ports:
         - protocol: TCP
           port: 80
+      name: AllowFromClient
+    - action: Drop
   egress:
     - action: Allow
       to:
         - podSelector:
             matchLabels:
-              app: dns
+              app: db
+      ports:
+        - protocol: TCP
+          port: 3306
+      name: AllowToDB
+    - action: Drop
+```
+
+## Antrea L7 NetworkPolicy Implementation
+
+In addition to layer 3 and layer 4 policies mentioned above, [Antrea Layer 7 NetworkPolicy](../antrea-l7-network-policy.md)
+is also supported in Antrea. The main difference is that Antrea layer 7 NetworkPolicy uses layer 7 protocol to match
+traffic, not layer 3 or layer 4 protocol.
+
+Consider the following Antrea layer 7 NetworkPolicy in the Application tier as an example for the remainder of this
+document.
+
+This ACNP is applied to all Pods with the label `app: client` in all Namespaces. For these Pods, egress DNS traffic on
+port 53 and egress HTTP traffic with host `*.bar.com` is allowed.
+
+```yaml
+apiVersion: crd.antrea.io/v1beta1
+kind: ClusterNetworkPolicy
+metadata:
+  name: allow-web-access-to-internal-domain
+spec:
+  priority: 6
+  tier: application
+  appliedTo:
+    - podSelector:
+        matchLabels:
+          app: client
+  egress:
+    - name: AllowDNS
+      action: Allow
       ports:
+        - protocol: TCP
+          port: 53
         - protocol: UDP
           port: 53
+    - name: AllowClientHTTP
+      action: Allow
+      l7Protocols:
+        - http:
+            host: "*.bar.com"
 ```
 
-This ACNP is applied to all Pods with the `app: server` label in all
-Namespaces. For these Pods, it drops TCP traffic on port 80 from all
-Pods which have the `app: notClient` label. In addition to the ingress rules,
-this policy also allows egress UDP traffic on port 53 to all Pods with the
-label `app: dns`. Similar to K8s NetworkPolicy, Antrea will only install OVS
-flows for this ACNP on Nodes for which some of the Pods are the target of the
-policy. Thus, we have scheduled three Pods (appServer, appDns, appNotClient)
-on the same Node and they have the following IP addresses:
+This ACNP is applied to all Pods with the label `app: web` in all Namespaces. For these Pods, only ingress HTTP
+traffic with method `GET` path `/api/v2/*` from the Pods with the label `app: client` is allowed, and others will
+be dropped.
 
-- appServer: 10.10.1.6
-- appNotClient: 10.10.1.7
-- appDns: 10.10.1.8
+```yaml
+apiVersion: crd.antrea.io/v1beta1
+kind: NetworkPolicy
```

Comment:
The text above says "This ACNP" but this policy is of type ANNP

Additionally, I would expect all the examples given in this document to "co-exist", but I am having trouble with this policy. Unless I am missing something this policy (`ingress-allow-http-request-to-api-v2`) will never actually be used, because `web-app-db-network-policy` will already allow all TCP:80 traffic from client Pods.

Reply:
> The text above says "This ACNP" but this policy is of type ANNP

This kind should be `ClusterNetworkPolicy`.

> I would expect all the examples given in this document to "co-exist"

Second to this. I am always expecting that the example flow can co-exist.

I have updated the policy `ingress-allow-http-request-to-api-v2`. It now matches ingress HTTP traffic on port 8080.

Reply:
I don't think that will work as long as you have a catch-all ingress `Drop` rule in `web-app-db-network-policy`, which has higher priority than `ingress-allow-http-request-to-api-v2`.

```yaml
  ingress:
    - action: Allow
      from:
        - podSelector:
            matchLabels:
              app: client
      ports:
        - protocol: TCP
          port: 80
      name: AllowFromClient
    - action: Drop
```

As per the current version of `web-app-db-network-policy`, all ingress traffic that is not TCP:80 traffic from a Pod labelled `app: client` will be dropped.

I cannot think of a great solution. It may make sense to have a different `app` that we can apply the policy to:

```yaml
apiVersion: crd.antrea.io/v1beta1
kind: ClusterNetworkPolicy
metadata:
  name: ingress-allow-http-request-to-api-v2
spec:
  priority: 6
  tier: application
  appliedTo:
    - podSelector:
        matchLabels:
          app: api
  ingress:
    - name: AllowFromClientL7
      action: Allow
      from:
        - podSelector:
            matchLabels:
              app: client
      ports:
        - protocol: TCP
          port: 8080
      l7Protocols:
        - http:
            path: "/api/v2/*"
            method: "GET"
    - action: Drop
```

The difference here is that we apply the policy to `app: api` Pods instead of `app: web`. **Alternatively, you could choose a higher priority for this priority (set it to 4), and drop the `action: Drop` catch-all rule in this policy.** You can keep the Drop catch-all rule in `web-app-db-network-policy`.

Reply:
> I don't think that will work as long as you have a catch-all ingress Drop rule in web-app-db-network-policy, which has higher priority than ingress-allow-http-request-to-api-v2.

You reminded me.  

> Alternatively, you could choose a higher priority for this priority (set it to 4), and drop the action: Drop catch-all rule in this policy. You can keep the Drop catch-all rule in web-app-db-network-policy.

This is a good idea! I'll update the document like this.